### ISSUE TITLE
DatabaseStore breaks transaction because it controls flow via exception

### ISSUE THREAD

**Issue by mailopl** [laravel/framework#33934]
- Laravel Version: 7.25.0
- PHP Version: 7.4.8
- Database Driver & Version: Postgres 12.1

### Description:

tl;dr
```
(in) --> Controller --> CommandBus --> TransactionMiddleware --> CommandHandler --> RateLimiter --> CacheRepository --> DatabaseStore
```
TransactionMiddleware opens a transaction on "db" connection. DatabaseManager uses same connection, to create DatabaseStore. DatabaseStore in add method (1), tries to insert a record (2) and it fails (3), breaking DB transaction. Exception is handled within DatabaseStore (3), that tries to update a record, using same connection, with aborted transaction.
Perhaps this was done for performance reasons, but in my case I just happened to be using connection with open transaction.

So I wanted to propose little refactoring that goes like this (unless I simply shouldn't use this on a connection with open transaction):
```
(1)
try {
    return $this->table()->insert(compact('key', 'value', 'expiration')); (2)
} catch (QueryException $e) {
    return $this->table() (3)
        ->where('key', $key)
        ->where('expiration', '<=', $this->getTime())
        ->update([
            'value' => $value,
            'expiration' => $expiration,
        ]) >= 1;
}
```
(if check, instead of exception control flow)
```
$item = $this->get($key);
if (null !== $item) {
    return $this->table()
            ->where('key', $key)
            ->where('expiration', '<=', $this->getTime())
            ->update([
                'value' => $value,
                'expiration' => $expiration,
            ]) >= 1;
}

return $this->table()->insert(compact('key', 'value', 'expiration'));
```

### Steps To Reproduce:
Call RateLimiter::hit() twice with same key, in a transaction, using underneath Illuminate\Cache\DatabaseStore.

PS. If this is not considered an issue, sorry for wasting your time. ☮️

**Comment by sisve** [comment]
Could you expand on how this is breaking any transactions?

**Comment by mailopl** [comment]
@sisve 
Sorry, should've been more precise.
I'm using Tactician's CommandBus, that utilizes simple TransactionMiddleware.

```
class TransactionMiddleware implements \League\Tactician\Middleware
{
    /**
     * @throws \Throwable
     */
    public function execute(object $command, callable $next)
    {
        try {
            \DB::beginTransaction();
            $returnValue = $next($command);
            \DB::commit();
        } catch (\Exception $e) {
            \DB::rollback();
            throw $e;
        }

        return $returnValue;
    }
}
```

One of my command handlers uses RateLimiter during transaction.
During that transaction, an exception is thrown because entry already exists as explained above. 
In the end, exception is caught on PHP side, but on DB side transaction is aborted.

**Comment by taylorotwell** [comment]
What are you wanting us to fix here?

**Comment by mailopl** [comment]
I was thinking something like, refactor this:

```
try {
    return $this->table()->insert(compact('key', 'value', 'expiration'));
} catch (QueryException $e) {
    return $this->table()
        ->where('key', $key)
        ->where('expiration', '<=', $this->getTime())
        ->update([
            'value' => $value,
            'expiration' => $expiration,
        ]) >= 1;
}
```
into +-  this:
```
$item = $this->get($key);
if (null !== $item) {
    return $this->table()
            ->where('key', $key)
            ->where('expiration', '<=', $this->getTime())
            ->update([
                'value' => $value,
                'expiration' => $expiration,
            ]) >= 1;
}

return $this->table()->insert(compact('key', 'value', 'expiration'));
```

**Comment by taylorotwell** [comment]
That introduces a race condition between the get and the insert / update.

**Comment by taylorotwell** [comment]
I'm not sure there is anything for us to change in the lock / rate limiter code. It is atomic by design.

**Comment by blakethepatton** [comment]
Maybe this could be updated to utilize upsert which is implemented in quite a few db languages? Pgsql has a nice [wiki page](https://wiki.postgresql.org/wiki/UPSERT#UPSERT_as_implemented_in_practice) outlining the syntax in the various languages. Not sure how many different languages laravel/eloquent supports but it'd be a fantastic feature to have available just in general especially with updateOrCreate. I'm sure I'm not the only one who gets the occasional message coming from bug tracking software about a record already existing and the only insert is inside a updateOrCreate.
